### PR TITLE
update e2e test for federation replicaset controlelr

### PR DESCRIPTION
**Pull request by jianhuiz**
e2e test to verify replicases synced to underlying clusters.

@quinton-hoole @nikhiljindal @deepak-vij @kshafiee @mwielgus

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31904)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

**Comment by nikhiljindal** [comment]
cc @kubernetes/sig-cluster-federation

File: test/e2e/federation-replicaset.go [hunk 1]
```diff
@@ -64,22 +71,129 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
 		})
 
 	})
+
+	// e2e cases for federation replicaset controller
```
Comment (quinton-hoole) [inline test/e2e/federation-replicaset.go]: nit: federated replicaset

File: test/e2e/federation-replicaset.go [hunk 2]
```diff
@@ -64,22 +71,129 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+	err := wait.Poll(10*time.Second, FederatedReplicaSetTimeout, func() (bool, error) {
+		frs, err := c.ReplicaSets(namespace).Get(replicaSetName)
+		if err != nil {
+			return false, err
+		}
+		specReplicas, statusReplicas := int32(0), int32(0)
+		for _, cluster := range clusters {
+			rs, err := cluster.ReplicaSets(namespace).Get(replicaSetName)
+			if err != nil && !errors.IsNotFound(err) {
+				By(fmt.Sprintf("Filed getting replicaset: %q/%q/%q, err: %v", cluster.name, namespace, replicaSetName, err))
```
Comment (quinton-hoole) [inline test/e2e/federation-replicaset.go]: nit: Failed

**Comment by jianhuiz** [comment]
comments addressed

**Comment by quinton-hoole** [comment]
@jianhuiz One more thing.  We don't seem to have any tests for the cluster weighting/min/max stuff.  Could you add those in a followup PR?

**Comment by quinton-hoole** [comment]
Code LGTM.  I assume that you have run these successfully @jianhuiz ?

File: test/e2e/federation-replicaset.go [hunk 3]
```diff
@@ -64,22 +71,129 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+		var (
+			clusters       map[string]*cluster
+			federationName string
+		)
+		BeforeEach(func() {
+			framework.SkipUnlessFederated(f.Client)
+			if federationName = os.Getenv("FEDERATION_NAME"); federationName == "" {
+				federationName = DefaultFederationName
+			}
+			registerClusters(clusters, UserAgentName, federationName, f)
```
Comment (nikhiljindal) [inline test/e2e/federation-replicaset.go]: clusters is nil here.
This test is failing with:

```
Test Panicked
    assignment to entry in nil map
    /usr/local/go/src/runtime/panic.go:443

    Full Stack Trace
    	/usr/local/go/src/runtime/panic.go:443 +0x4e9
    k8s.io/kubernetes/test/e2e.registerClusters(0x0, 0x34c90e0, 0xe, 0x34c90d0, 0xa, 0xc820381040, 0x0, 0x0)
    	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:207 +0xbc2
    k8s.io/kubernetes/test/e2e.glob.func32.2.1()
    	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-replicaset.go:86 +0xf0
```

Need to add `clusters = map[string]*cluster{}`
Am sending a fix
cc @jianhuiz

**Pull request merged**